### ISSUE TITLE
Add direct access to the input of date picker

### ISSUE THREAD

**Issue by AdrianoRuberto** [ant-design/ant-design#18743]
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
The `onChange` function and `format` are triggered only when the date is "finished", thus it's impossible to add a mask on the input change (or to use any third-party library like react-text-mask to format the input while the use is typing). One solution is to add the possibility to use a custom input component in DatePicker.

It's not a a duplicate of https://github.com/ant-design/ant-design/pull/17366 because even if the format props is a function, it will not trigger as the user type.

### What does the proposed API look like?
```jsx
<DatePicker
   InputComponent={(props) => <CustomInput {...props} />}
>
```
The props would be an object with the props taken by the current input.

**Comment by afc163** [comment]
https://github.com/ant-design/ant-design/issues/18376#issuecomment-523281731

**Comment by AdrianoRuberto** [comment]
How this
> [#18376 (comment)](https://github.com/ant-design/ant-design/issues/18376#issuecomment-523281731)

is relevant ? It only hide the whole component and do not give access to the DatePicker's input. What I meant by "direct access to the input" is to have the complete control on the input of the DatePicker component. To have the DatePicker (with the calendar below the input), but with a custom input component.

**Comment by afc163** [comment]
It is a workaround for custom tigger, we may not provider built-in custom input for DatePicker in short term due to lack of maintainers.

**Comment by lkenny97** [comment]
I needed to reformat the date while the user types it in. This is my workaround for people who may face this problem in the future.

I created a ref using useRef hook and assigned it to my DatePicker. I then wrote this code in my useEffect:

`const reformatDate = (event) => {
    const strippedInput = event.target.value.replaceAll("-", "")
    let newInput = "";
    for (let i = 0; i < strippedInput.length; i += 1) {
      if (i === 2 || i === 4)
        newInput += "-"
      newInput += strippedInput.charAt(i)
    }
    event.target.value = newInput
  }

  React.useEffect(() => {

    ReactDOM.findDOMNode(datePickerRef.current).children[0].children[0]
      .addEventListener("keyup", reformatDate)

    return () => {
      ReactDOM.findDOMNode(datePickerRef.current).children[0].children[0].removeEventListener("keyup", reformatDate)
    }
  }, [])`

I know that this is not the most efficient code but it solved my problem for now. Please add an onInput or onType or something like that to DatePicker.

**Comment by nlukestik** [comment]
Hi! Pardon my ignorance, but how do you created the ref and assigned it to DatePicker component? Since it's not possible to  pass ref to a component and I'm not being able to do it with forwardRef either..
Thanks in advance!

**Comment by hpedrorodrigues** [comment]
Here is a workaround (with [imask][imask]) that applies a mask and reformats the date while users type it in to avoid the need to add `/`/`-` when typing the date.

```typescript
import React from 'react'
import {
  DatePicker as AntdDatePicker,
  DatePickerProps as AntDatePickerProps,
} from 'antd'
import IMask from 'imask'
import moment from 'moment'

const DATE_FORMAT = 'MM/DD/YYYY'
const MASKED = IMask.createMask({
  blocks: {
    DD: {from: 1, mask: IMask.MaskedRange, to: 31},
    MM: {from: 1, mask: IMask.MaskedRange, to: 12},
    YYYY: {from: 1900, mask: IMask.MaskedRange, to: Number.MAX_VALUE},
  },
  format: (date: Date) => moment(date).format(DATE_FORMAT),
  mask: Date,
  parse: (date: string) => moment(date, DATE_FORMAT),
  pattern: DATE_FORMAT,
})

export type DatePickerProps = Omit<AntDatePickerProps, 'format' | 'picker' | 'onKeyDown'>

const DatePicker = (props: DatePickerProps) => (
  <AntdDatePicker
    format={DATE_FORMAT}
    onKeyDown={(event: React.KeyboardEvent<HTMLInputElement>) => {
      const input = event.target as HTMLInputElement
      input.value = MASKED.resolve(input.value)
    }}
    picker="date"
    placeholder={DATE_FORMAT.toLowerCase()}
    {...props}
  />
)

export default DatePicker
```

[imask]: https://imask.js.org

**Comment by darfilp** [comment]
> Hi! Pardon my ignorance, but how do you created the ref and assigned it to DatePicker component? Since it's not possible to pass ref to a component and I'm not being able to do it with forwardRef either.. Thanks in advance!

Everything worked out for me, but when editing the date breaks, how can I fix it?

**Comment by coder054** [comment]
````
inputRender={(a) => {
  const { onChange, ...rest } = a
  return (
    <input
      {...rest}
      onChange={(e) => {
        console.log('aaa e', e)
        return onChange && onChange(e)
      }}
    />
  )
}}
````
